pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -28,42 +32,67 @@ import (
 // Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
 // will be turned into one or more actual node instances within the cluster after bin packing.
 type Node struct {
+	Name                string
 	Constraints         *v1alpha5.Constraints
 	InstanceTypeOptions []cloudprovider.InstanceType
 	Pods                []*v1.Pod
 
 	requests v1.ResourceList
+	client   client.Client
 }
 
-func NewNode(constraints *v1alpha5.Constraints, daemonResources v1.ResourceList, instanceTypes []cloudprovider.InstanceType) *Node {
-	return &Node{
+var nodeID int64
+
+func NewNode(c client.Client, constraints *v1alpha5.Constraints, daemonResources v1.ResourceList, instanceTypes []cloudprovider.InstanceType) *Node {
+	n := &Node{
+		Name:                fmt.Sprintf("node-%04d", atomic.AddInt64(&nodeID, 1)),
 		Constraints:         constraints.DeepCopy(),
 		InstanceTypeOptions: instanceTypes,
 		requests:            daemonResources,
+		client:              c,
 	}
+
+	n.Constraints.Requirements = n.Constraints.Requirements.Add(v1.NodeSelectorRequirement{
+		Key:      v1.LabelHostname,
+		Operator: v1.NodeSelectorOpIn,
+		Values:   []string{n.Name},
+	})
+	return n
 }
 
-func (n *Node) Add(pod *v1.Pod) error {
+func (n *Node) addPod(pod *v1.Pod, commit bool) error {
```

Comment:
It looks like `commit` here is used to show distinguish between a function call that is meant to actually bind the pod to the node but the other just checks constraints. Can you separate this function into one which checks the constraints, and one which binds the pod? It seems weird to me to add a boolean to distinguish when a function will actually do something vs when it's doing a check.

Reply:
I reworked it a bit, maybe somewhat cleaner.